subsystem/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+
+use logging::log;
+
+/// Top-level subsystem manager.
+///
+/// An application is composed of a number of long-lived subsystems. The [Manager] type starts
+/// and manages the life cycle of the subsystems. Whenever a subsystem exits, all other subsystems
+/// are requested to terminate and the manager is shut down.
+pub struct Manager {
+    // Manager name
+    name: &'static str,
+
+    // Order all subsystems to shutdown
+    shutdown_request_tx: broadcast::Sender<()>,
+    shutdown_request_rx: broadcast::Receiver<()>,
+
+    // Signal a subsystem is shutting down. This is taken as an instruction for all subsystems to
+    // shut down. Shutdown completion is detected by all senders having closed this channel.
+    shutting_down_tx: mpsc::Sender<()>,
+    shutting_down_rx: mpsc::Receiver<()>,
+}
+
+impl Manager {
+    /// Initialise a new subsystem manager.
+    pub fn new(name: &'static str) -> Self {
+        log::info!("Initialising subsystem manager {}", name);
+
+        let (shutdown_request_tx, shutdown_request_rx) = broadcast::channel(1);
+        let (shutting_down_tx, shutting_down_rx) = mpsc::channel(1);
+
+        Self {
+            name,
+            shutdown_request_tx,
+            shutdown_request_rx,
+            shutting_down_tx,
+            shutting_down_rx,
+        }
+    }
+
+    /// Start a subsystem.
+    ///
+    /// A subsystem has to handle shutdown and call requests. It can also react to external IO
+    /// events. If the subsystem handles *only* calls and shutdown requests without interaction
+    /// with any additional IO, use the [Manager::start_passive] convenience method.
+    ///
+    /// A typical skeleton of a subsystem looks like this:
+    /// ```no_run
+    /// # let manager = subsystem::Manager::new("app");
+    /// let subsystem = manager.start("my-subsystem", |mut call, mut shutdown| async move {
+    ///     loop {
+    ///         tokio::select! {
+    ///             // Shutdown received, break out of the loop.
+    ///             () = shutdown.recv() => { break; }
+    ///             // Handle calls. An object representing the subsystem is passed in.
+    ///             func = call.recv() => { func(todo!("put an argument here")); }
+    ///             // Handle any other IO events here
+    ///         };
+    ///     }
+    /// });
+    /// # let _ = subsystem.call(|()| ());  // Fix the call type to avoid amnbiguity.
+    /// ```
+    pub fn start<T: 'static + Send, F: 'static + Send + Future<Output = ()>>(
+        &self,
+        name: &'static str,
+        subsystem: impl 'static + Send + FnOnce(CallRequest<T>, ShutdownRequest) -> F,
+    ) -> Subsystem<T> {
+        // App name
+        let appname = self.name;
+        // Shutdown-related channels
+        let shutting_down_tx = self.shutting_down_tx.clone();
+        let shutdown_rq = ShutdownRequest(self.shutdown_request_tx.subscribe());
+        // Call related channels
+        let (action_tx, action_rx) = mpsc::channel(64);
```

Comment:
Can we make this 64 a parameter? I think there should be a configuration struct for everything related (?)

Reply:
Aye this could be done

Reply:
I was thinking of using that, but thought it's a little of an over kill... better than nothing though.